File: app/src/main/java/com/soundcloud/android/discovery/SelectionItemAdapter.java
package com.soundcloud.android.discovery;

import com.google.auto.factory.AutoFactory;
import com.google.auto.factory.Provided;
import com.soundcloud.android.discovery.DiscoveryCardViewModel.MultipleContentSelectionCard;
import com.soundcloud.android.model.Urn;
import com.soundcloud.android.presentation.RecyclerItemAdapter;
import com.soundcloud.java.optional.Optional;
import io.reactivex.subjects.PublishSubject;

import android.support.v7.widget.RecyclerView;
import android.view.View;

@AutoFactory(allowSubclasses = true)
class SelectionItemAdapter extends RecyclerItemAdapter<SelectionItemViewModel, RecyclerView.ViewHolder> {

    private Optional<Urn> selectionUrn = Optional.absent();

    SelectionItemAdapter(@Provided SelectionItemRendererFactory selectionItemRendererFactory, PublishSubject<SelectionItemViewModel> selectionItemClickListener) {
        super(selectionItemRendererFactory.create(selectionItemClickListener));
    }

    public Optional<Urn> selectionUrn() {
        return selectionUrn;
    }

    void updateSelection(MultipleContentSelectionCard selection) {
        selectionUrn = Optional.of(selection.getSelectionUrn());
        clear();
        onNext(selection.getSelectionItems());
    }

    @Override
    protected ViewHolder createViewHolder(View itemView) {
        return new ViewHolder(itemView);
    }

    @Override
    public int getBasicItemViewType(int position) {
        return 0;
    }
}

File: app/src/main/java/com/soundcloud/android/discovery/SelectionItemAdapter.kt
package com.soundcloud.android.discovery

import android.support.v7.widget.RecyclerView
import android.view.View
import com.soundcloud.android.discovery.DiscoveryCardViewModel.MultipleContentSelectionCard
import com.soundcloud.android.model.Urn
import com.soundcloud.android.presentation.RecyclerItemAdapter
import com.soundcloud.android.utils.OpenForTesting
import io.reactivex.subjects.PublishSubject
import javax.inject.Inject

@OpenForTesting
class SelectionItemAdapter(selectionItemRendererFactory: SelectionItemRenderer.Factory,
                           selectionItemClickListener: PublishSubject<SelectionItemViewModel>,
                           var selectionUrn: Urn? = null) :
        RecyclerItemAdapter<SelectionItemViewModel, RecyclerView.ViewHolder>(selectionItemRendererFactory.create(selectionItemClickListener)) {

    fun updateSelection(selection: MultipleContentSelectionCard) {
        selectionUrn = selection.selectionUrn
        clear()
        onNext(selection.selectionItems)
    }

    override fun createViewHolder(itemView: View) = RecyclerItemAdapter.ViewHolder(itemView)

    override fun getBasicItemViewType(position: Int): Int = 0

    @OpenForTesting
    data class Factory
    @Inject
    constructor(private val selectionItemRendererFactory: SelectionItemRenderer.Factory) {
        fun create(selectionItemClickListener: PublishSubject<SelectionItemViewModel>) = SelectionItemAdapter(selectionItemRendererFactory, selectionItemClickListener)
    }
}
